internal/pkg/deploy/cloudformation/env.go

```diff
@@ -159,67 +155,49 @@ func (cf CloudFormation) UpgradeEnvironment(in *deploy.CreateEnvironmentInput) e
-	return cf.upgradeEnvironment(in, func(param *awscfn.Parameter) *awscfn.Parameter {
-		if aws.StringValue(param.ParameterKey) == includeLoadBalancerParamKey {
-			// "IncludePublicLoadBalancer" has been deprecated in favor of "ALBWorkloads".
-			// We need to populate this parameter so that the env ALB is not deleted.
+	return cf.upgradeEnvironment(in, func(new *awscfn.Parameter, old *awscfn.Parameter) *awscfn.Parameter {
+		if new == nil {
+			return nil
+		}
+		if aws.StringValue(new.ParameterKey) == stack.EnvParamALBWorkloadsKey {
 			return &awscfn.Parameter{
 				ParameterKey:   aws.String(stack.EnvParamALBWorkloadsKey),
 				ParameterValue: aws.String(strings.Join(lbWebServices, ",")),
 			}
 		}
+		if old == nil {
+			return new
+		}
 		return &awscfn.Parameter{
-			ParameterKey:     param.ParameterKey,
+			ParameterKey:     new.ParameterKey,
 			UsePreviousValue: aws.Bool(true),
 		}
 	})
 }
 
-func (cf CloudFormation) upgradeEnvironment(in *deploy.CreateEnvironmentInput, transformParam func(param *awscfn.Parameter) *awscfn.Parameter) error {
-	bucketARN, err := arn.Parse(in.ArtifactBucketARN)
-	if err != nil {
-		return err
-	}
-	stackConfig := stack.NewEnvStackConfig(in)
-	url, err := cf.uploadStackTemplateToS3(bucketARN.Resource, stackConfig)
-	if err != nil {
-		return err
-	}
-	s, err := toStackFromS3(stackConfig, url)
+func (cf CloudFormation) upgradeEnvironment(in *deploy.CreateEnvironmentInput, transformParam func(new *awscfn.Parameter, old *awscfn.Parameter) *awscfn.Parameter) error {
+	s, err := cf.environmentStack(in)
 	if err != nil {
 		return err
 	}
 
 	for {
-		descr, err := cf.cfnClient.Describe(s.Name)
+		var (
+			descr *cloudformation.StackDescription
+			err   error
+		)
+		descr, err = cf.waitAndDescribeStack(s.Name)
 		if err != nil {
-			return fmt.Errorf("describe stack %s: %w", s.Name, err)
-		}
-
-		if cloudformation.StackStatus(aws.StringValue(descr.StackStatus)).InProgress() {
-			// There is already an update happening to the environment stack.
-			// Best-effort try to wait for the existing update to be over before retrying.
-			_ = cf.cfnClient.WaitForUpdate(context.Background(), s.Name)
-			continue
+			return err
 		}
 
-		// Remove params that only exist in old template; use previous values for params that
-		// exist in both old and new template; use new values for params that only exist in new template.
-		paramSet := make(map[string]*awscfn.Parameter)
-		for _, param := range s.Parameters {
-			paramSet[aws.StringValue(param.ParameterKey)] = param
-		}
-		for _, param := range descr.Parameters {
-			param = transformParam(param)
-			paramKey := aws.StringValue(param.ParameterKey)
-			if _, ok := paramSet[paramKey]; !ok {
-				continue
-			}
-			paramSet[paramKey] = param
-		}
-		var params []*awscfn.Parameter
-		for _, param := range paramSet {
-			params = append(params, param)
+		// Generally, we want to:
```

Comment:
nit: maybe drop the "Generally"
```suggestion
		// We want to:
```

Reply:
Or maybe just 
```suggestion
```
and move the comment to `transformParameters`